### PR TITLE
Dls 6716 pack at business address base page

### PR DESCRIPTION
**Pull request by GinnyThomas**
Pack at business page - no logic for routing to page - will be handled in Navigator. 
Ability to keep and recall subscription added - tests for this will be included on ticket specific to this work.

### CODE REVIEW

File: app/controllers/PackAtBusinessAddressController.scala [hunk 1]
```diff
@@ -46,21 +47,23 @@ class PackAtBusinessAddressController @Inject()(
...
-
-      val preparedForm = request.userAnswers.get(PackAtBusinessAddressPage) match {
-        case None => form
-        case Some(value) => form.fill(value)
-      }
-
-      Ok(view(preparedForm, mode))
+      val businessName = request.subscription.orgName
+      val businessAddress = request.subscription.address
+        lazy val preparedForm = request.userAnswers.get(PackAtBusinessAddressPage) match {
```
Comment (RaminEsfandiari) [inline app/controllers/PackAtBusinessAddressController.scala]: Could have this at the same level of the two above vals. So tab back once maybe?

File: app/controllers/PackAtBusinessAddressController.scala [hunk 2]
```diff
@@ -46,21 +47,23 @@ class PackAtBusinessAddressController @Inject()(
...
 
   def onSubmit(mode: Mode): Action[AnyContent] = (identify andThen getData andThen requireData).async {
     implicit request =>
+      val businessName = request.subscription.orgName
+      val businessAddress = request.subscription.address
 
       form.bindFromRequest().fold(
         formWithErrors =>
-          Future.successful(BadRequest(view(formWithErrors, mode))),
+          Future.successful(BadRequest(view(formWithErrors, businessName: String, businessAddress: UkAddress, mode))),
```
Comment (RaminEsfandiari) [inline app/controllers/PackAtBusinessAddressController.scala]: Don't really need the types on businessName and businessAddress but it is best to either declare types for all arguments or for none (these are things that can be agreed as part of team coding style as long as it doesn't break any HMRC wide style rules of course)

File: app/controllers/actions/IdentifierAction.scala [hunk 3]
```diff
@@ -45,16 +46,28 @@ class AuthenticatedIdentifierAction @Inject()(
...
 
     authorised(AuthProviders(GovernmentGateway)).retrieve(allEnrolments) { enrolments =>
-      (getSdilEnrolment(enrolments), getUtr(enrolments)) match {
-        case (Some(e), _) => block(IdentifierRequest(request, e.value))
-        case (None, Some(utr)) =>  sdilConnector.retrieveSubscription(utr, "utr").flatMap {
-          case Some(subscription) =>
-            sdilConnector.oldestPendingReturnPeriod(utr).flatMap { returnPeriod =>
-              block(IdentifierRequest(request, EnrolmentIdentifier("sdil", subscription.sdilRef).value,List.empty,returnPeriod))
+      val maybeUtr = getUtr(enrolments)
+      val maybeSdil = getSdilEnrolment(enrolments)
```
Comment (RaminEsfandiari) [inline app/controllers/actions/IdentifierAction.scala:50]: Not sure about these maybe prefixes. I understand why but just don't think it is necessary nor clean. No need to change as you didn't name these though :)

File: app/models/requests/IdentifierRequest.scala [hunk 4]
```diff
@@ -16,7 +16,12 @@
...
-import models.{SmallProducer,ReturnPeriod}
+import models.retrieved.RetrievedSubscription
+import models.{ReturnPeriod, SmallProducer}
 import play.api.mvc.{Request, WrappedRequest}
 
-case class IdentifierRequest[A] (request: Request[A], sdilEnrolment: String, smallProducerList :List[SmallProducer] = List.empty, returnPeriod: Option[ReturnPeriod] = None) extends WrappedRequest[A](request)
+case class IdentifierRequest[A] (request: Request[A],
+                                 sdilEnrolment: String,
+                                 subscription: RetrievedSubscription,
+                                 //smallProducerList :List[SmallProducer] = List.empty,
```
Comment (RaminEsfandiari) [inline app/models/requests/IdentifierRequest.scala]: Remove this line please, we can always refer back to commit history if we need something from the past.

File: test/controllers/PackAtBusinessAddressControllerSpec.scala [hunk 5]
```diff
@@ -108,6 +118,8 @@ class PackAtBusinessAddressControllerSpec extends SpecBase with MockitoSugar {
         val result = route(application, request).value
 
         status(result) mustEqual SEE_OTHER
+        redirectLocation(result).value mustEqual onwardRoute.url
```
Comment (RaminEsfandiari) [inline test/controllers/PackAtBusinessAddressControllerSpec.scala]: What part of the code is this covering? If I where to break something that this test line would catch what would it be?

File: test/controllers/PackAtBusinessAddressControllerSpec.scala [hunk 6]
```diff
@@ -129,9 +141,10 @@ class PackAtBusinessAddressControllerSpec extends SpecBase with MockitoSugar {
...
 
         status(result) mustEqual BAD_REQUEST
-        contentAsString(result) mustEqual view(boundForm, NormalMode)(request, messages(application)).toString
-//        page.getElementsContainingText(usersRetrievedSubscription.orgName).toString == true
-//        page.getElementsContainingText(usersRetrievedSubscription.address.toString).`val`() == true
+        contentAsString(result) mustEqual view(boundForm, businessName, businessAddress, NormalMode)(request, messages(application)).toString
+
+        page.getElementsContainingText(usersRetrievedSubscription.orgName).toString == true
+        page.getElementsContainingText(usersRetrievedSubscription.address.toString).`val`() == true
 //        page.getElementsByTag("a").text() must include(Messages("packAtBusinessAddress.required"))
```
Comment (RaminEsfandiari) [inline test/controllers/PackAtBusinessAddressControllerSpec.scala]: You can remove this line

File: test/controllers/actions/SessionActionSpec.scala [hunk 7]
```diff
@@ -14,64 +14,80 @@
...
-          val request = FakeRequest().withSession(SessionKeys.sessionId -> "foo")
-
-          val result = controller.onPageLoad()(request)
-          status(result) mustBe OK
-        }
-      }
-    }
-  }
-}
+///*
```
Comment (RaminEsfandiari) [inline test/controllers/actions/SessionActionSpec.scala]: Why are these commented out?

**Review by RaminEsfandiari — CHANGES_REQUESTED** [review]
Looks good overall but a few minor comments to address.

File: test/controllers/PackAtBusinessAddressControllerSpec.scala (hunk 5 above)
Comment (GinnyThomas) [reply test/controllers/PackAtBusinessAddressControllerSpec.scala]: left in by lack of attention.  Removed as it has no function.

File: test/controllers/actions/SessionActionSpec.scala (hunk 7 above)
Comment (GinnyThomas) [reply test/controllers/actions/SessionActionSpec.scala]: session action no longer being used.  File deleted.

File: test/controllers/PackAtBusinessAddressControllerSpec.scala [hunk 8]
```diff
@@ -20,31 +20,34 @@ import base.SpecBase
...
 import org.jsoup.Jsoup
 import play.api.i18n.Messages
-import play.api.i18n.Messages.implicitMessagesProviderToMessages
+import play.api.mvc.Call
 
 import scala.concurrent.Future
 
 class PackAtBusinessAddressControllerSpec extends SpecBase with MockitoSugar {
-
+  def onwardRoute = Call("GET", "#")
```
Comment (RaminEsfandiari) [inline test/controllers/PackAtBusinessAddressControllerSpec.scala]: Don you need this?

File: test/controllers/PackAtBusinessAddressControllerSpec.scala [hunk 9]
```diff
@@ -96,6 +105,7 @@ class PackAtBusinessAddressControllerSpec extends SpecBase with MockitoSugar {
       val application =
         applicationBuilder(userAnswers = Some(emptyUserAnswers))
           .overrides(
+//            bind[Navigator].toInstance(new FakeNavigator(onwardRoute)),
```
Comment (RaminEsfandiari) [inline test/controllers/PackAtBusinessAddressControllerSpec.scala]: Could you remove this please?

**Review by RaminEsfandiari — CHANGES_REQUESTED** [review]
Just a couple of lines to clean up please

File: test/controllers/PackAtBusinessAddressControllerSpec.scala (hunk 9 above)
Comment (GinnyThomas) [reply test/controllers/PackAtBusinessAddressControllerSpec.scala]: no.  Thanks for spotting it

**Review by RaminEsfandiari — APPROVED** [review]

**Pull request merged**